# TASK 6.1 SOURCING OPEN DATA

**This notebook contains the following scripts:**
    
    1. Import libraries
    2. Import data set
    3. Data Cleaning
    4. Export Cleaned Data Set

# 1. Import libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

# 2. Import data set

In [2]:
#define path
path = r'C:\Users\ddgal\OneDrive\Desktop\Career Foundry Tasks\Data Immersion\Achievement 6'

In [3]:
#import data set as df and remove index column
df = pd.read_csv(os.path.join(path, 'Original Data', 'Customer-Churn-Records.csv'), index_col=0)

In [4]:
#check shape of data set
df.shape

(10000, 17)

We have 10,000 rows and 18 columns.

In [5]:
#check columns
df.head()

CustomerId   Surname  CreditScore Geography  Gender  Age  Tenure  \
RowNumber                                                                     
1            15634602  Hargrave          619    France  Female   42       2   
2            15647311      Hill          608     Spain  Female   41       1   
3            15619304      Onio          502    France  Female   42       8   
4            15701354      Boni          699    France  Female   39       1   
5            15737888  Mitchell          850     Spain  Female   43       2   

             Balance  NumOfProducts  HasCrCard  IsActiveMember  \
RowNumber                                                        
1               0.00              1          1               1   
2           83807.86              1          0               1   
3          159660.80              3          1               0   
4               0.00              2          0               0   
5          125510.82              1          1               1   

           EstimatedSalary  Exited  Complain  Satisfaction Score Card Type  \
RowNumber                                                                    
1                101348.88       1         1                   2   DIAMOND   
2                112542.58       0         1                   3   DIAMOND   
3                113931.57       1         1                   3   DIAMOND   
4                 93826.63       0         0                   5      GOLD   
5                 79084.10       0         0                   5      GOLD   

           Point Earned  
RowNumber                
1                   464  
2                   456  
3                   377  
4                   350  
5                   425

# 3. Data Cleaning

**Missing Values**

In [6]:
#check for missing values
df.isnull().sum()

CustomerId            0
Surname               0
CreditScore           0
Geography             0
Gender                0
Age                   0
Tenure                0
Balance               0
NumOfProducts         0
HasCrCard             0
IsActiveMember        0
EstimatedSalary       0
Exited                0
Complain              0
Satisfaction Score    0
Card Type             0
Point Earned          0
dtype: int64

There are no blanks in this data sets.

**Duplicates**

In [7]:
#check for duplicates 
df_dups = df[df.duplicated()]
df_dups

Empty DataFrame
Columns: [CustomerId, Surname, CreditScore, Geography, Gender, Age, Tenure, Balance, NumOfProducts, HasCrCard, IsActiveMember, EstimatedSalary, Exited, Complain, Satisfaction Score, Card Type, Point Earned]
Index: []

There are no duplicates in this data set.

**Mixed-Type Variables**

In [8]:
#check for mixed-typed data
for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

There is no mixed-type data.

**Variable Types**

In [9]:
#check variable types

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 1 to 10000
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   CustomerId          10000 non-null  int64  
 1   Surname             10000 non-null  object 
 2   CreditScore         10000 non-null  int64  
 3   Geography           10000 non-null  object 
 4   Gender              10000 non-null  object 
 5   Age                 10000 non-null  int64  
 6   Tenure              10000 non-null  int64  
 7   Balance             10000 non-null  float64
 8   NumOfProducts       10000 non-null  int64  
 9   HasCrCard           10000 non-null  int64  
 10  IsActiveMember      10000 non-null  int64  
 11  EstimatedSalary     10000 non-null  float64
 12  Exited              10000 non-null  int64  
 13  Complain            10000 non-null  int64  
 14  Satisfaction Score  10000 non-null  int64  
 15  Card Type           10000 non-null  object 
 16  Poin

The columns HasCrCard, IsActiveMember, Exited, and complain should be strings since 1 represents yes and 0 represents no.  I will replace the 1s with yes and 0s with no.  This will make any visualizations made in the future a lot easier to understand. 

The replacement was successfull since the four columns have 10000 non-null entries and are now objects.

**Basic Statistics**

In [10]:
#check basic statistical information
df.describe()

CustomerId   CreditScore           Age        Tenure        Balance  \
count  1.000000e+04  10000.000000  10000.000000  10000.000000   10000.000000   
mean   1.569094e+07    650.528800     38.921800      5.012800   76485.889288   
std    7.193619e+04     96.653299     10.487806      2.892174   62397.405202   
min    1.556570e+07    350.000000     18.000000      0.000000       0.000000   
25%    1.562853e+07    584.000000     32.000000      3.000000       0.000000   
50%    1.569074e+07    652.000000     37.000000      5.000000   97198.540000   
75%    1.575323e+07    718.000000     44.000000      7.000000  127644.240000   
max    1.581569e+07    850.000000     92.000000     10.000000  250898.090000   

       NumOfProducts    HasCrCard  IsActiveMember  EstimatedSalary  \
count   10000.000000  10000.00000    10000.000000     10000.000000   
mean        1.530200      0.70550        0.515100    100090.239881   
std         0.581654      0.45584        0.499797     57510.492818   
min         1.000000      0.00000        0.000000        11.580000   
25%         1.000000      0.00000        0.000000     51002.110000   
50%         1.000000      1.00000        1.000000    100193.915000   
75%         2.000000      1.00000        1.000000    149388.247500   
max         4.000000      1.00000        1.000000    199992.480000   

             Exited      Complain  Satisfaction Score  Point Earned  
count  10000.000000  10000.000000        10000.000000  10000.000000  
mean       0.203800      0.204400            3.013800    606.515100  
std        0.402842      0.403283            1.405919    225.924839  
min        0.000000      0.000000            1.000000    119.000000  
25%        0.000000      0.000000            2.000000    410.000000  
50%        0.000000      0.000000            3.000000    605.000000  
75%        0.000000      0.000000            4.000000    801.000000  
max        1.000000      1.000000            5.000000   1000.000000

Everything looks normal except for Estimated Salary.   The min estimated Salary is $11.58 which makes no sense since it is highly unlikely that in Europe someone has such a low Salary. Further analysis is required on this issue.

In [11]:
#check how many values are under Estimated Salary $15000 which is based on Spain min wage.  Spain has the lowest min wage of the three countries.

df[df['EstimatedSalary'] < 15000]

CustomerId    Surname  CreditScore Geography  Gender  Age  Tenure  \
RowNumber                                                                      
7            15592531   Bartlett          822    France    Male   50       7   
17           15737452      Romeo          653   Germany    Male   58       1   
18           15788218  Henderson          549     Spain  Female   24       9   
24           15725737     Mosman          669    France    Male   46       3   
59           15623944      T'ien          511     Spain  Female   66       4   
...               ...        ...          ...       ...     ...  ...     ...   
9949         15735078  Onwughara          724   Germany  Female   53       1   
9955         15739850   Trentino          645    France    Male   45       6   
9975         15695474     Barker          583    France    Male   33       7   
9980         15692664     Diribe          677    France  Female   58       1   
9988         15588839    Mancini          606     Spain    Male   30       8   

             Balance  NumOfProducts  HasCrCard  IsActiveMember  \
RowNumber                                                        
7               0.00              2          1               1   
17         132602.88              1          1               0   
18              0.00              2          1               1   
24              0.00              2          0               1   
59              0.00              1          1               0   
...              ...            ...        ...             ...   
9949       139687.66              2          1               1   
9955       155417.61              1          0               1   
9975       122531.86              1          1               0   
9980        90022.85              1          0               1   
9988       180307.73              2          1               1   

           EstimatedSalary  Exited  Complain  Satisfaction Score Card Type  \
RowNumber                                                                    
7                 10062.80       0         0                   2    SILVER   
17                 5097.67       1         0                   2    SILVER   
18                14406.41       0         0                   3    SILVER   
24                 8487.75       0         0                   2    SILVER   
59                 1643.11       1         1                   5    SILVER   
...                    ...     ...       ...                 ...       ...   
9949              12913.92       0         0                   2   DIAMOND   
9955               3449.22       0         0                   3  PLATINUM   
9975              13549.24       0         0                   1    SILVER   
9980               2988.28       0         0                   4      GOLD   
9988               1914.41       0         0                   3    SILVER   

           Point Earned  
RowNumber                
7                   206  
17                  163  
18                  544  
24                  665  
59                  549  
...                 ...  
9949                371  
9955                810  
9975                313  
9980                355  
9988                567  

[750 rows x 17 columns]

So there are 750 rows with Estimated Salary of less than $15000 which is 7.5% of the data.  This is will only affect insights derived from the Estimated Salary, so for now I will leave it. Possible solutions are to find the average minimum wage for the three countries and replace anything lower than that with Nan or to remove those entire rows only when deriving insights from Estimated Salary.  The best solution might be to create a new variable which divides the Estimated Salary into groups such as Low-income, 

**Drop Columns**

In [12]:
#drop columns Surname since it has PII and check

df_clean = df.drop(columns = ['Surname'])
df_clean.shape

(10000, 16)

Column drop was successful since there is one less column.

**Renaming Columns**

In [13]:
#rename columns for overall consistency
df_clean_2 = df_clean.rename(columns = {'CustomerId' : 'Customer_id', 'CreditScore' : 'Credit_score', 
                               'NumOfProducts' : 'Number_of_products', 'HasCrCard' : 'Has_credit_card',
                               'IsActiveMember' : 'Is_active_member', 'EstimatedSalary' : 'Estimated_salary',
                               'Satisfaction Score' : 'Satisfaction_score', 'Card Type' : 'Card_type',
                               'Point Earned' : 'Points_earned'})

In [14]:
#check df_clean_3
df_clean_2.head()

Customer_id  Credit_score Geography  Gender  Age  Tenure  \
RowNumber                                                             
1             15634602           619    France  Female   42       2   
2             15647311           608     Spain  Female   41       1   
3             15619304           502    France  Female   42       8   
4             15701354           699    France  Female   39       1   
5             15737888           850     Spain  Female   43       2   

             Balance  Number_of_products  Has_credit_card  Is_active_member  \
RowNumber                                                                     
1               0.00                   1                1                 1   
2           83807.86                   1                0                 1   
3          159660.80                   3                1                 0   
4               0.00                   2                0                 0   
5          125510.82                   1                1                 1   

           Estimated_salary  Exited  Complain  Satisfaction_score Card_type  \
RowNumber                                                                     
1                 101348.88       1         1                   2   DIAMOND   
2                 112542.58       0         1                   3   DIAMOND   
3                 113931.57       1         1                   3   DIAMOND   
4                  93826.63       0         0                   5      GOLD   
5                  79084.10       0         0                   5      GOLD   

           Points_earned  
RowNumber                 
1                    464  
2                    456  
3                    377  
4                    350  
5                    425

All the column renaming was executed successfully.

# 4. Export Cleaned Data Set

In [15]:
#export df_clean_3

df_clean_2.to_csv(os.path.join(path, 'Prepared Data', 'cust_churn_records_clean.csv'))